test/fw/ptl/utils/pbs_crayutils.py

```diff
@@ -0,0 +1,145 @@
+        """
+        Run apstat command on cray/craysim and parse its output
+
+        :param options: options to pass to apstat command
+        :type options: str
+        :returns: tuple of (node status, node summary)
+        """
+        status = []
+        summary = {}
+        count = 0
+        options = '-rn'
+        hostname = socket.gethostname()
+        platform = self.du.get_platform(hostname)
+        apstat_env = os.environ
+        apstat_cmd = "apstat"
+        if 'cray' not in platform:
+            return (status, summary)
+        if 'craysim' in platform:
+            lib_path = '$LD_LIBRARY_PATH:/opt/alps/tester/usr/lib/'
+            apstat_env['LD_LIBRARY_PATH'] = lib_path
+            apstat_env['ALPS_CONFIG_FILE'] = '/opt/alps/tester/alps.conf'
+            apstat_env['apsched_sharedDir'] = '/opt/alps/tester/'
+            apstat_cmd = "/opt/alps/tester/usr/bin/apstat -d ."
+        cmd_run = self.du.run_cmd(hostname, [apstat_cmd, options],
+                                  as_script=True, wait_on_script=True,
+                                  env=apstat_env)
+        cmd_result = cmd_run['out']
+        keys = cmd_result[0].split()
+        # Add a key 'Mode' because 'State' is composed of two list items, e.g:
+        # State = 'UP  B', where Mode = 'B'
+        k2 = ['Mode']
+        keys = keys[0:3] + k2 + keys[3:]
+        cmd_iter = iter(cmd_result)
+        for line in cmd_iter:
+            if count == 0:
+                count = 1
+                continue
+            if "Compute node summary" in line:
+                summary_line = next(cmd_iter)
+                summary_keys = summary_line.split()
+                summary_data = next(cmd_iter).split()
+                sum_index = 0
+                for a in summary_keys:
+                    summary[a] = summary_data[sum_index]
+                    sum_index += 1
+                break
+            index = 0
+            obj = {}
+            line = line.split()
+            for i in range(0, len(line)):
+                obj[keys[index]] = line[i]
+                if keys[index] == 'State':
+                    obj[keys[index]] = line[i] + "  " + line[i + 1]
+                index += 1
+            # If there is no Apids in the apstat then use 'None' as the value
+            if "Apids" in obj:
+                pass
+            else:
+                obj["Apids"] = None
+            status.append(obj)
+        return (status, summary)
+
+    def count_node_in_state(self, cnsumm='up'):
+        """
+        Return the value of any one of the following parameters as shown in
+        the 'Compute Node Summary' section of 'apstat -rn' output:
+        arch, config, up, resv, use, avail, down
+
+        :param cnsumm: parameter which is being queried, defaults to 'up'
+        :type cnsumm: str
+        :returns: value of parameter being queried
+        """
+        return int(self.node_summary[cnsumm])
+
+    def count_node_State(self, state='UP  B'):
```

Comment:
With this new function in place, do we need "count_node_in_state" method as well? I think both these methods are doing the same thing but their source is different. In count_node_in_state() we collect this information from node summary but in this new method we compute it using node status.
If you happen to remove "count_node_in_state" then I'd suggest also get rid of node_summary and how it is populated because I do not see it will be useful anywhere then.

Reply:
We still need "count_node_in_state" and I have renamed it "count_node_summ" to refer to the summary section that is being used.  It is still used in lots of places in the tests where the summary information is sufficient.